### PR TITLE
Adding & enhancing client apis over gRPC

### PR DESCRIPTION
**Pull request by amanbha**
Summary of Changes:
- Removed individual clients for Publish, State and Invoke.
- Added single DaprClient and its implementation which is a wrapper over auto generated gRPC code.
- Added client calls for InvokeMethod, State, Binding, Publish and secret which work over gRPC
- Aded Etag and state options support to state apis.
- Added metadata support to Apis.
- Added a DaprClientBuilder to create DaprClient.
- Renamed autogenerated grpc client project.
- Updated unit tests to use gRPC.

It will close multiple issues: 
https://github.com/dapr/dotnet-sdk/issues/226
https://github.com/dapr/dotnet-sdk/issues/225
https://github.com/dapr/dotnet-sdk/issues/224
https://github.com/dapr/dotnet-sdk/issues/223
https://github.com/dapr/dotnet-sdk/issues/243
https://github.com/dapr/dotnet-sdk/issues/221
https://github.com/dapr/dotnet-sdk/issues/142

### CODE REVIEW

File: all.sln [hunk 1]
```diff
@@ -41,7 +41,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ActorClient", "samples\Acto
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "AspNetCore", "AspNetCore", "{DCBEAB71-0BDE-42F4-8187-F576942CB038}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Dapr.Client.Grpc", "src\Dapr.Client.Grpc\Dapr.Client.Grpc.csproj", "{BCE5A721-69C5-4510-97ED-931ADE587466}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Dapr.Client.Autogen.Grpc", "src\Dapr.Client.Autogen.Grpc\Dapr.Client.Autogen.Grpc.csproj", "{BCE5A721-69C5-4510-97ED-931ADE587466}"
```
Comment (rynowak) [inline all.sln]: What's the role of this as a separate project now that we're taking a dependency on it? Should this and `Dapr.Client` just be one project?

Comment (amanbha) [reply all.sln]: The raw autogenerated client code would still be available as a separate package if users want to consume it directly.

File: src/Dapr.Client/DaprClient.cs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        /// <param name="storeName">The name of the state store.</param>
+        /// <param name="key">The state key.</param>
+        /// <param name="consistencyMode">The consistency mode <see cref="ConsistencyMode" />.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> that can be used to cancel the operation.</param>
+        /// <typeparam name="TValue">The data type of the value to read.</typeparam>
+        /// <returns>A <see cref="ValueTask" /> that will return the <see cref="StateEntry{T}" /> when the operation has completed.</returns>
+        public async ValueTask<StateEntry<TValue>> GetStateEntryAsync<TValue>(string storeName, string key, ConsistencyMode? consistencyMode = default, CancellationToken cancellationToken = default)
+        {
+            storeName.ThrowIfNullOrEmpty(nameof(storeName));
+            key.ThrowIfNullOrEmpty(nameof(key));
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/DaprClient.cs]: :) Should it just be an helper method then?

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+    using Grpc.Core;
+    using Grpc.Net.Client;
+    using Autogenerated = Dapr.Client.Autogen.Grpc;
+
+    /// <summary>
+    /// A client for interacting with the Dapr endpoints.
+    /// </summary>
+    internal class DaprClientGrpc : DaprClient
+    {
+        private readonly Autogenerated.Dapr.DaprClient client;
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/DaprClientGrpc.cs]: yes, thats the type of autogenerated clients

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+            this.jsonSerializerOptions = jsonSerializerOptions;
+            this.client = new Autogenerated.Dapr.DaprClient(channel);
+        }
+
+        #region Publish Apis
+        /// <inheritdoc/>
+        public override Task PublishEventAsync<TRequest>(string topicName, TRequest publishContent, CancellationToken cancellationToken = default)
+        {
+            topicName.ThrowIfNullOrEmpty(nameof(topicName));
+            publishContent.ThrowIfNull(nameof(publishContent));
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/DaprClientGrpc.cs]: The other overload allows to publish without any cotnent.

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+            }
+
+            var response = await this.MakeGrpcCallHandleError(
+                 (options) =>
+                 {
+                     return client.GetSecretAsync(envelope, options);
+                 },
+                 cancellationToken);
+
+            return response.Data.ToDictionary(kv => kv.Key, kv => kv.Value);
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/DaprClientGrpc.cs]: yes, they are case sensitive.

File: src/Dapr.Client/ETag.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Dapr.Client
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public class ETag : IEquatable<ETag>
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/ETag.cs]: I will remove this struct and use string for an etag

File: src/Dapr.Client/ETag.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        /// <summary>
+        /// Gets the value of ETag.
+        /// </summary>
+        public string Value { get; }
+
+        /// <inheritdoc/>
+        public bool Equals([AllowNull] ETag other)
+        {
+            return this.Value.Equals(other.Value);
+        }
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/ETag.cs]: I am making Etag just a string.

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 8]
```diff
@@ -0,0 +1,622 @@
...
+            await stream.FlushAsync();
+
+            // set the position to beginning of stream.
+            stream.Seek(0, SeekOrigin.Begin);
+
+            return new Any
+            {
+                Value = await ByteString.FromStreamAsync(stream)
+            };
+        }
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/DaprClientGrpc.cs]: Now directly using JsonSerializer.SerializeToUtf8Bytes

File: src/Dapr.Client/RetryOptions.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+        /// <summary>
+        /// Retry pattern, can be either linear or exponential.
+        /// </summary>
+        public RetryMode? RetryMode { get; set; }
+
+        /// <summary>
+        /// Maximum number of retries.
+        /// </summary>
+        public int? RetryThreshold { get; set; }
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/RetryOptions.cs]: Will add sensible defaults in next release, for this release I am adding apis.

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 10]
```diff
@@ -0,0 +1,622 @@
...
+            key.ThrowIfNullOrEmpty(nameof(key));
+
+            try
+            {
+                await this.MakeSaveStateCallAsync<TValue>(storeName, key, value, etag, stateOptions, metadata, cancellationToken);
+                return true;
+            }
+            catch (Exception)
+            {
+                // log?
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/DaprClientGrpc.cs]: Will consider adding logging only when the whole class logs other things.

File: src/Dapr.AspNetCore/Dapr.AspNetCore.csproj [hunk 11]
```diff
@@ -13,6 +13,10 @@
     <FrameworkReference Include="Microsoft.AspNetCore.App" />
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Grpc.Net.ClientFactory" Version="2.27.0" />
```
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.AspNetCore/Dapr.AspNetCore.csproj]: Removed,.

File: src/Dapr.Client/DaprClientBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        string daprEndpoint;
+        JsonSerializerOptions jsonSerializerOptions;
+        GrpcChannelOptions gRPCChannelOptions;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DaprClientBuilder"/> class.
+        /// </summary>
+        public DaprClientBuilder()
+        {
+            var defaultPort = Environment.GetEnvironmentVariable("DAPR_GRPC_PORT") ?? "52918";
```
Comment (JamesNK) [inline src/Dapr.Client/DaprClientBuilder.cs]: Make 52918 a constant? Or add a comment explaining why this default.

File: src/Dapr.Client/DaprClientBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+        /// <summary>
+        /// Overrides the default endpoint used by IDaprClient for conencting to Dapr runtime.
+        /// </summary>
+        /// <param name="daprEndpoint">Endpoint to use for making calls to Dapr runtime. 
+        /// Default endpoint used is http://127.0.0.1:DAPR_GRPC_PORT.</param>
+        /// <returns>DaprClientBuilder instance.</returns>
+        public DaprClientBuilder UseEndpoint(string daprEndpoint)
+        {
+            this.daprEndpoint = daprEndpoint;
```
Comment (JamesNK) [inline src/Dapr.Client/DaprClientBuilder.cs]: What happens if you set the endpoint to null?

File: src/Dapr.Client/DaprClientBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+            GrpcChannel channel;
+            if (this.gRPCChannelOptions == null)
+            {
+                channel = GrpcChannel.ForAddress(this.daprEndpoint);
+            }
+            else 
+            {
+                channel = GrpcChannel.ForAddress(this.daprEndpoint, this.gRPCChannelOptions);
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Dapr.Client/DaprClientBuilder.cs]: Could do:

```
channel = GrpcChannel.ForAddress(this.daprEndpoint, this.gRPCChannelOptions ?? new GrpcChannelOptions());
```

File: src/Dapr.Client/StateOptions.cs [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    using System.Collections.Generic;
+    using System.Text;
+
+    /// <summary>
+    /// Options when perfroming state operations with Dapr.
+    /// </summary>
+    public class StateOptions
+    {
+        /// <summary>
+        ///
```
Comment (JamesNK) [inline src/Dapr.Client/StateOptions.cs]: Docs?

File: src/Dapr.Client/Dapr.Client.csproj [hunk 16]
```diff
@@ -8,5 +8,14 @@
   <PropertyGroup>
     <Description>This package contains the reference assemblies for developing services using Dapr.</Description>
   </PropertyGroup>
+  <ItemGroup>
```
Comment (JamesNK) [inline src/Dapr.Client/Dapr.Client.csproj]: Dapr.Client.csproj targets netcoreapp3.1

Have you considered netstandard2.1?

File: src/Dapr.Client/DaprClientGrpc.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Dapr.Client/DaprClientGrpc.cs]: You have to use ByteString for byte[]. ByteString is designed to be immutable. It takes a copy of any bytes you pass to it.

Yeah, its not good for allocations.

When the large serialization improvements have landed in Protobuf I'm going to see about adding a `ByteString.FromBytesUnsafe(byte[])` that doesn't take a copy when creating a ByteString.

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 17]
```diff
@@ -0,0 +1,622 @@
...
+    using Google.Protobuf;
+    using Google.Protobuf.WellKnownTypes;
+    using Grpc.Core;
+    using Grpc.Net.Client;
+    using Autogenerated = Dapr.Client.Autogen.Grpc;
+
+    /// <summary>
+    /// A client for interacting with the Dapr endpoints.
+    /// </summary>
+    internal class DaprClientGrpc : DaprClient
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Dapr.Client/DaprClientGrpc.cs]: Looks fine. It's all unary calls. Very simple!

File: src/Dapr.Client/Dapr.Client.csproj (hunk 16 above)
Comment (amanbha) [reply src/Dapr.Client/Dapr.Client.csproj]: Thats the decision we made early in the project to sue .netcore 3.*. https://github.com/dapr/dotnet-sdk/issues/15

File: samples/AspNetCore/ControllerSample/Controllers/SampleController.cs [hunk 18]
```diff
@@ -40,13 +41,13 @@ public ActionResult<Account> Get([FromState(StoreName)]StateEntry<Account> accou
         /// Method for depositing to account as specified in transaction.
         /// </summary>
         /// <param name="transaction">Transaction info.</param>
-        /// <param name="stateClient">State client to interact with Dapr runtime.</param>
+        /// <param name="daprClient">State client to interact with Dapr runtime.</param>
         /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
         [Topic("deposit")]
         [HttpPost("deposit")]
-        public async Task<ActionResult<Account>> Deposit(Transaction transaction, [FromServices] StateClient stateClient)
+        public async Task<ActionResult<Account>> Deposit(Transaction transaction, [FromServices] DaprClient daprClient)
```
[the remark this answers is not in the document]
Comment (amanbha) [reply samples/AspNetCore/ControllerSample/Controllers/SampleController.cs]: Nothing specific, just following the pattern which existed before these changes.

File: src/Dapr.Client/Dapr.Client.csproj (hunk 16 above)
Comment (JamesNK) [reply src/Dapr.Client/Dapr.Client.csproj]: Ok.

Normally I'd say a client library should target netstandard2.x to make it usable by the most people, but given how dapr is used, it is unlikely anyone will ever call a dapr service directly from a Xamarin/Unity/UWP app 😄

File: src/Dapr.Client/DaprClient.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Client/DaprClient.cs]: Feel free to do what you like, but it's pretty surprising to see extension methods used for argument validation compared with:

- A non-extension static method
- just writing the `if` 

Extension methods are pretty hard to discover if you're not using an IDE, and many beginners to C# would be surprised that an extension method is value to call on a null value.

File: samples/Client/DaprClient/Program.cs [hunk 19]
```diff
@@ -82,10 +82,31 @@ internal static async Task InvokeMethodAsync(DaprClient client)
             Console.WriteLine("Deleted State!");
         }
 
+        // This example shows how to invoke a method on a service listening on http.
+        // In such a scenario, a key value pair of http.verb and the verb must be added to the metadata parameter.
+        internal static async Task InvokeMethodOnHttpServiceAsync(DaprClient client)
+        {
+            MyData data = new MyData() { Message = "mydata" };
+            Dictionary<string, string> m = new Dictionary<string, string>();
+            m.Add("http.verb", "POST");
```
Comment (rynowak) [inline samples/Client/DaprClient/Program.cs]: Isn't POST the default for invoking methods? It's a wierd sample to just show that it's possible to do what dapr does by default 😆

File: src/Dapr.Client/DaprClient.cs [hunk 20]
```diff
@@ -59,6 +59,13 @@ public abstract class DaprClient
         /// <param name="metadata">A key/value pair to pass to the method to invoke.</param>
         /// <param name="cancellationToken">A <see cref="CancellationToken" /> that can be used to cancel the operation.</param>
         /// <returns>A <see cref="Task" /> that will complete when the operation has completed.</returns>
+        /// <remarks>If the target service is listening on http, a key value pair of http.verb and the verb must be added to the metadata parameter.
+        /// For example:
```
Comment (rynowak) [inline src/Dapr.Client/DaprClient.cs]: `<example>`

File: src/Dapr.Client/DaprClient.cs [hunk 21]
```diff
@@ -223,7 +226,7 @@ public async ValueTask<StateEntry<TValue>> GetStateEntryAsync<TValue>(string sto
         public abstract ValueTask<bool> TryDeleteStateAsync(
             string storeName,
             string key,
-            ETag etag,
+            string etag,
```
Comment (rynowak) [inline src/Dapr.Client/DaprClient.cs]: I don't have a big objection to this, but the rationale for making ETag its own type is to make it clear to users that they **should not** try to guess a good value for an etag. Making it a simple string makes it appear like something users can create.

File: src/Dapr.Client/RetryOptions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Client/RetryOptions.cs]: So what will happen? Throw? Format your harddrive? Hack the planet?

File: src/Dapr.Client/Dapr.Client.csproj (hunk 16 above)
Comment (rynowak) [reply src/Dapr.Client/Dapr.Client.csproj]: Yeah we went through all of this already. Dapr is for server side apps, and it's a new-ish project so there's no legacy.

File: samples/Client/DaprClient/Program.cs (hunk 19 above)
Comment (amanbha) [reply samples/Client/DaprClient/Program.cs]: yes, will change it to use GET verb and mention that POST is default.

File: src/Dapr.Client/DaprClient.cs (hunk 21 above)
Comment (amanbha) [reply src/Dapr.Client/DaprClient.cs]: I agree with the rationale, will add it back and make ti a struct.

File: src/Dapr.Client/DaprClient.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (amanbha) [reply src/Dapr.Client/DaprClient.cs]: I want community to contribute even beginners, so will stick to the norm and use a static method.

**Comment by amanbha** [comment]
@rynowak Could you please approve it again, addressing some of the review comments resetted the review.

**Comment by amanbha** [comment]
@LMWF  @JamesNK  Could you please approve it as well.

**Pull request merged**